### DISCUSSION TITLE
Map matching issues

### DISCUSSION THREAD

**Discussion by nilsnolde** [valhalla/valhalla#3524] [General]
This is more of a general plea for the sake of valhalla's reputation and I won't link all the issues regarding map matching here, but there's tons of open ones. I don't have the (unpaid) bandwidth to tackle any of those at the moment, nor is our company very much using map-matching which could also justify me working on it. There'll be [a presentation](https://pretalx.com/fossgis2022/talk/PVXDJD/) at German FOSSGIS conference early March, so I'll definitely get some feedback from Siavash on his experience, it's rare to know of such a gigantic, applied use case.

Can the issues be summarized here eventually? There's so many, it's hard to keep track of, but maybe someone (wink wink) kept a mental model of the main underlying problem(s)?

Also there doesn't seem to be a commercial player out there, e.g. mapbox or us, who's interested enough to resolve those issues. So I'd like to get some feedback if there's common interest to fund a crowd-funding campaign so we can eventually resolve them. If there's enough :+1: 's on this thread, I'd like to start one after consulting with more experienced people how much effort it'd be. I wouldn't estimate any less than $10k as a gut feeling, so I assume individual pledges would need to be considerable.

**Comment by kevinkreiser** [comment]
@nilsnolde let me answer some of the parts you mentioned

> This is more of a general plea for the sake of valhalla's reputation

I don't think it hurts the projects reputation that not everything is perfect and there isnt a huge team of engineers behind it pushing on it 24/7. If people come here expecting that, their expectations are wrong and its not our job to fulfill them :smile: This is a common problem that faces many open source projects and it's ok to do, as you have done many times, some education that what people want from us isnt always our top priority :smile: 

> but there's tons of open ones

you and i both know that in many cases spurious issues can and will be opened due to lack of understanding or additional debugging on the reporters part. anyway even if we go with the assumption that most of them are legitimate, the vast majority of those will be issues where the map matching has a discontinuity (a break in the match), this can be due to:

- the underlying data (go fix osm)
- assumptions we make in the costing model (pass one of the ignore options)
- the user inputting too narrow a radius on a given point (increase your radius 3x, this is what mapbox does)

I suspect if we had the time to actually look at every single report we'd find that one of these 3 factors would fix 50% of the "bug" reports.

> Can the issues be summarized here eventually

A quick search over github issues over the last 1.5 years (ie. after the big refactor we did earlier reports would not be relevant) shows:

- edge index for interpolated points is wrong in trace_attributes (serialization problem)
- location index is wrong (serialization problem)
- that bug you found where walk or snap doesnt fallback correctly but just snap does work (probably partial state needs cleared or something)
- 2 questions about bad map matching results (probably stuff like i was mentioning above but could also be corner cases with the map matching algorithm)

> I'd like to start one (crowd-funding campaign) after consulting with more experienced people how much effort it'd be

Let me start with the second part first. We shouldnt expect Mapbox (I was the guy at Mapbox doing the map matching work anyway) or someone else to swoop in and maintain parts of the project or that we need to raise money to get things fixed. In other words, its not that we dont have the experience to maintain our own software :smile: (the bugs above arent even map matching perse) its that we dont have the time to get to the bottom of every single github issue that is posted or review every PR that comes in. At the end of the day there are very few of us and very many people on the internet. But i guess you know that.

Reading between the lines I think I understand that you would be interested in crowd funding your work on Valhalla; basically giving people a way to fund your time to prioritize their issue. That sounds fine to me, feel free to post a link to your crowd funding page, but please keep it to one place so that we don't have advertisements all over the repo and please don't post dollar amounts here.

And one more thing I want to stress: people should understand that crowdfunding options are not the primary recourse to fixing/prioritizing their issues. Valhalla will remain, as it always has been, a free and open source project where everyone is welcome to contribute, whether that is posting debugging findings, adding tests, adding docs, testing features or writing PRs in general. We want to fix the issues that people encounter but gauranteeing anything regarding timelines is difficult when you are talking about volunteer work.

**Comment by ImreSamu** [comment]
> Also there doesn't seem to be a commercial player out there, 

Is it possible to create a map-matching related "Google Summer of Code 2022" project?
( not for the bugfixing .. but for the long-term future of Valhalla )
 
The timeline: https://developers.google.com/open-source/gsoc/timeline
* "February 7 - 18:00 UTC"  - Mentoring organizations can begin submitting applications to Google
* "February 21 - 18:00 UTC" - Mentoring organization application deadline

------
One of my biggest problems:  no simple scripting API  for using Valhalla "Internals" 
- writing plug-ins/extensions ( like the PostgreSQL extensions )
- or just using for debugging the data.

dreaming:  creating a Valhalla "alternative" map-matching/routing/etc plug-in in https://julialang.org  and using GPU for fast processing. 
   * Julia is [very promising machine learning language](https://discourse.julialang.org/t/state-of-machine-learning-in-julia/74385) ;  [HN thread](https://news.ycombinator.com/item?id=29902846)
   * And Map-matching + machine Learning / GPU  is a hot research topic.
      *  https://scholar.google.com/scholar?q=map+matching+deep+learning
      *  https://scholar.google.com/scholar?q=map+matching+gpu
